BUG: longdouble(int) does not work

eric-wieser opened the pull request:
Fixes gh-9968

Needs tests yet, which I can't write without access to a platform with `longdouble` having higher precision than `double`.

eric-wieser committed:
BUG: longdouble(int) does not work

Fixes gh-9968

eric-wieser commented:
Rebased. This code has a deeply worrying heisenbug in MSVC...

eric-wieser commented on numpy/core/src/common/npy_longdouble.c:
```diff
@@ -94,3 +95,77 @@ npy_longdouble_to_PyLong(npy_longdouble ldval)
+    }
+
+    // Without this line, MSVC produces garbage (optimizes out result!?)
+    printf("Double form is %f\n", (double) result);
```
I kid you not. Without this, `np.longdouble(0)` returns `-104892023.0`.

eric-wieser replied:
In even more frightening news, replacing this line with:
```
// Without this line, MSVC produces garbage (optimizes out result!?)
printf("abc\n");
return result;
```

Causes `np.longdouble(x)` to always return `4.0`, which is `len('abc\n')`.

mattip replied:
could the compiler output from [godbot](https://godbolt.org/) help work out what is going on?

tylerjereddy replied:
This appears to be because the function prototype was not available before the function was called in `arraytypes.c.src`, which caused just as much chaos on my Mac until I added the appropriate `include` there, which I've now pushed in with my initial adjustments.

I suggest we might benefit from `warning: implicit declaration of function` into a compiler error--looks like this warning is effectively understating a break in C99 standard.

tylerjereddy committed:
MAINT: reviewer adjustments in PR 10723

* add unit test for gh-9968, which currently
fails on master

* npy_longdouble.h include was missing in
arraytypes.c.src, causing all sorts of
unstable behavior for calls to
npy_longdouble_from_PyLong()

* correct control flow for PyBool_Type in
string_to_long_double() function; include
a unit test for np.longdouble(bool)

tylerjereddy commented on numpy/core/getlimits.py:
```diff
@@ -156,8 +156,8 @@ def _register_known_types():
     ld = ntypes.longdouble
-    epsneg_f128 = exp2(ld(-113))
-    tiny_f128 = exp2(ld(-16382))
+    epsneg_f128 = 2 ** -113.0
```
@eric-wieser can I actually do this? I also noticed we have `exp2l` buried somewhere--regardless `getlimits.py` is having issues with these exp2 calls involving long double.

eric-wieser replied:
No, this will underflow and produce 0 on all platforms if you don't cast an operand to longdouble

tylerjereddy replied:
Reverted -- no longer causes issues, thanks.

tylerjereddy commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -30,6 +30,7 @@
 #include <emmintrin.h>
 #endif
 
+#include "npy_longdouble.h"
```
without this, adding `printf` or even calling code multiple times caused stability issues; I think we should just make it an error if the function prototype isn't available

eric-wieser replied:
Good catch! Please do make this a compilation error!

tylerjereddy commented on numpy/core/src/common/npy_longdouble.c:
```diff
@@ -94,3 +95,88 @@ npy_longdouble_to_PyLong(npy_longdouble ldval)
+    / NOTE: the block below was causing an error in
+    / 10 tests in numpy/ma/tests/test_core.py:
+    / RuntimeError: Could not parse long as longdouble: True
+    */
```
Maybe need to adjust the code below for masked array handling?

tylerjereddy replied:
resolved with proper bool handling noted in commit message

tylerjereddy reviewed (commented):
There will still be a few unit test failures left in CI after my revision push, but hopefully this dislodges some of the blocking issues here--in particular, the nasty bug with the function prototype inclusion preventing sane debugging.

I'm expecting 3 test failures related to fpe underflow expectations & things in `getlimits.py`, which Eric can probably solve with more confidence than I.

tylerjereddy commented:
Yeah, just seeing 3 test failures on Windows / Mac now in CI, so this should hopefully be easier to get over the line now.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -328,6 +329,17 @@ string_to_long_double(PyObject*op)
+    if (PyLong_Check(op)
+#if !defined(NPY_PY3K)
+        || PyInt_Check(op)
+#endif
```
I think we just need to avoid running this code-path for booleans (and adding a test for `np.longdouble(True)`)

tylerjereddy replied:
yes! that worked! I guess because of bools in masked arrays or something like that!

tylerjereddy commented:
All tests pass locally & all conversations marked as resolved after applying Eric's revisions to my revisions. Fingers crossed for the CI to agree.

tylerjereddy commented:
Remaining CI failures can be summarized as:

1 test fails with the following on platforms where `np.longdouble==np.double`:
`RuntimeWarning: overflow encountered in conversion from python long`

A DEBUG build has a problem with `epsneg_f128 = exp2(ld(-113))`.
`untimeError: Could not parse long as longdouble:` followed by a bunch of strange characters

tylerjereddy committed:
MAINT: more adjustments in PR 10723

* adjust test_longdouble_from_int to properly
handle platforms where np.longdouble and
np.double have roughly equivalent limits

tylerjereddy commented on numpy/core/tests/test_longdouble.py:
```diff
@@ -205,3 +206,28 @@ def test_fromstring_foreign_sep(self):
+        # max longdouble value on all platforms
+        if np.allclose(np.finfo(np.longdouble).max,
+                       np.finfo(np.double).max) and w:
+            assert w[0].category is RuntimeWarning
```
This seemed to help on Windows in my hands. Basically, we check for a warning on platforms where longdouble simple isn't big enough to handle the test. It'll typically produce something like `inf==inf` for the big test case, so I think can avoid having to skip / xfail.

tylerjereddy commented on numpy/core/src/common/npy_longdouble.c:
```diff
@@ -94,3 +95,82 @@ npy_longdouble_to_PyLong(npy_longdouble ldval)
+        return -1;
+    }
+
+    /* Extra characters at the end of the string, or nothing parsed */
```
@eric-wieser could DEBUG mode be putting extra characters in here somewhere? I still need to debug that from the Travis fail we're seeing.

I couldn't reproduce on my mac earlier, but I'm not entirely clear on the right CFLAGS or whatever to induce the issue, and if the issue is portable to clang.

tylerjereddy reviewed (commented):
Revised to hopefully fix 1 test failure, although CI may be confused by issues discussed in #13448.

Still expecting DEBUG mode fail on Travis--discussed briefly in a review comment.

tylerjereddy commented:
One other thought here, since we don't see the Azure 32-bit failure (ran before the merge of #13449), is that Azure may be doing branch instead of merge tests?

eric-wieser commented:
> RuntimeError: Could not parse long as longdouble: followed by a bunch of strange characters

I think it's a lifetime problem, and I have a local fix for it. Will attempt to merge with your fixes at some point...

tylerjereddy commented:
Ok, that sounds good. I see that Azure is actually doing merges, so that's good too.

eric-wieser committed:
BUG: Fix incorrect reference counting of bytes

This needs to be kept alive so we can check the last character of the string

tylerjereddy commented:
Great, thanks Eric

Would be good if our line cov data eventually covers more than 1 platform, but the few "missing" error handling lines in the new C function were hit during the review / debugging of this PR at various stages anyway, so we can be pretty confident they are hit on Windows / Power8 in CI, etc.
